[5.0] [WIP] Math - Implement Generic Math interfaces

vovatrykoz opened the pull request:
This is my first PR to this repository, so I would appreciate any feedback, especially on the testing part!

### Purpose of this PR

The purpose of this PR is to add changes proposed in #1761. The issue proposes adding generic math interfaces introduced in .NET 7 to existing OpenTK Math structs.

At the time of writing this description, I have only added interfaces to the [Vector2](https://github.com/vovatrykoz/opentk/blob/implement-generic-math-interfaces/src/OpenTK.Mathematics/Vector/Vector2.cs). struct. I will add the interfaces to other structs, but I wanted to open a pull request to get early feedback.

It is not possible to implement some interfaces due to constraint on TSelf: TSelf needs to implement that same mathematical interface. For example, in [IAdditionOperators](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.iadditionoperators-3?view=net-9.0), TSelf also needs to implement IAdditionOperators<TSelf,TOther,TResult>. Therefore, the following interfaces have not been added, even if there are operators that could in theory work:
- IMultiplyOperators<float, Vector2, Vector2>
- IMultiplyOperators<Matrix2, Vector2, Vector2>
- IMultiplyOperators<Quaternion, Vector2, Vector2>

While it is not possible to do much about the floats, it is possible to make Matrix and Quaternion compatible with these interfaces, but the question is should we?

### Testing status

Originally I wanted to add unit tests that would check if a specific struct satisfies generic math constraints. However, I ran into some issues with regard to that. First of all, the compiler adds warning FS3536 when trying to test that. For example, the following code:
```
    [<Properties(Arbitrary = [| typeof<OpenTKGen> |])>]
    module ``Math Generics`` =
        //
        [<Property>]
        let ``Vector2 implements IAdditionOperators<Vector2, Vector2, Vector2> interface`` (v : Vector2) =
            let vectorType = v.GetType()
            let interfaceType = typeof<IAdditionOperators<Vector2, Vector2, Vector2>>

            let implementsInterface (vectorType: Type) (interfaceType: Type) =
                vectorType.IsAssignableTo(interfaceType)

            Assert.True(implementsInterface vectorType interfaceType)
```

Will produce this warning
![image](https://github.com/user-attachments/assets/ee64f322-de42-46be-b6b6-79bef2d4d64d)

According to the [F# documentation](https://github.com/fsharp/fslang-design/blob/main/FSharp-7.0/FS-1124-interfaces-with-static-abstract-members.md#warning-when-using-iwsams-as-types): "At all other locations the use of an IWSAM constraint results in the warning. Note this includes `typeof<ISomeInterface<...>>` - reflecting over IWSAM constraints is sufficiently rare that this is not given a special dispensation."

Some unit tests result in the FS0001 error, for reasons I do not fully understand. Here is another piece of code as an example
```
    [<Properties(Arbitrary = [| typeof<OpenTKGen> |])>]
    module ``Math Generics`` =
        //
        [<Property>]
        let ``Vector2 implements IMultiplyOperators<Vector2, Vector2, Vector2> interface`` (v : Vector2) =
            let vectorType = v.GetType()
            let interfaceType = typeof<IMultiplyOperators<Vector2, Vector2, Vector2>>

            let implementsInterface (vectorType: Type) (interfaceType: Type) =
                vectorType.IsAssignableTo(interfaceType)

            Assert.True(implementsInterface vectorType interfaceType)

        [<Property>]
        let ``Vector2 implements IMultiplyOperators<Vector2, float, Vector2> interface`` (v : Vector2) =
            let vectorType = v.GetType()
            let interfaceType = typeof<IMultiplyOperators<Vector2, float32, Vector2>>

            let implementsInterface (vectorType: Type) (interfaceType: Type) =
                vectorType.IsAssignableTo(interfaceType)

            Assert.True(implementsInterface vectorType interfaceType)
```

In here, the first unit test will result in the following error
![image](https://github.com/user-attachments/assets/037bb2af-25d1-4e0a-a595-810141cf06c4)

However, the second one will only have a warning
![image](https://github.com/user-attachments/assets/b245ed9a-53b5-4871-8b38-3b670126fa88)

I am not sure why F# has such a hard time understanding that these are valid types. This error pops up when trying to test the following interfaces:
- IMultiplyOperators<Vector2, Vector2, Vector2>
- IMultiplyOperators<Vector2, Matrix2, Vector2>
- IDivisionOperators<Vector2, Vector2, Vector2>

I would appreciate advice on how to properly test this or if this issue even needs unit testing. The only other way I can think of is simply typing out examples like so:

```
using System.Numerics;

Animation<Vector2> animation = new(Vector2.One, Vector2.One);

class Animation<T>(T value, T update) where T : IAdditionOperators<T, T, T>
{
    public T Update()
    {
        value += update;
        return value;
    }
}
```

This example can be found here, in Program.cs: https://github.com/vovatrykoz/OpenTk_test

vovatrykoz commented:
Also, there are other interfaces in the Generic Math that could be of interest, more specifically:
[IAdditiveIdentity](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.iadditiveidentity-2?view=net-9.0)
[IMinMaxValue](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.iminmaxvalue-1?view=net-9.0)
[IMultiplicativeIdentity](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.imultiplicativeidentity-2?view=net-9.0)

Should they also be added to the existing math types? For example, for IAdditiveIdentity the library has have Vector.Zero and Matrix.Zero. 
For IMinMaxValue there is PositiveInfinity and NegativeInfinity (for types working with doubles and floats)
For IMultiplicativeIdentity there is identity matrix and vectors consisting of ones (Vector.One)

vovatrykoz commented:
Ok, this should be it, I am not sure if it makes sense to add these interfaces to other math structs, since they only have == and != operators

NogginBops commented:
Great work!
I'll review this PR next time I work on 5.0 stuff (trying to get 4.9.0 out first).
Will likely add the interfaces you mention above to make the generic math support "complete".

vovatrykoz commented:
I forgot to mention that I discovered that Matrix3 does not have some of the operators that Matrix2 and Matrix4 have. More specifically, the following operators are not implemented for Matrix3

Matrix3 operator +(Matrix3 left, Matrix3 right)
Matrix3 operator -(Matrix3 left, Matrix3 right)
Matrix3 operator *(Matrix3 left, float right)
Matrix3 operator *(float left, Matrix3 right)

I could add these operators and add corresponding math interfaces where possible, but I was unsure if this will "pollute" the PR sin a way, since originally it was about adding interfaces only

NogginBops commented:
I think it would be fine to add these operators in this PR if that allows all of the matrix types to have the same interfaces.

vovatrykoz commented:
Ok, I just added the following operators and corresponding interfaces (where possible) to the [Matrix3](https://github.com/vovatrykoz/opentk/blob/implement-generic-math-interfaces/src/OpenTK.Mathematics/Matrix/Matrix3.cs) struct:
- Matrix3 operator +(Matrix3 left, Matrix3 right)
- Matrix3 operator -(Matrix3 left, Matrix3 right)
- Matrix3 operator *(Matrix3 left, float right)
- Matrix3 operator *(float left, Matrix3 right)
- Matrix3x4 operator *(Matrix3 left, Matrix3x4right)

NogginBops commented:
@vovatrykoz are you interested in adding the rest of the math functions needed to make for example `Matrix3d` have the same interfaces as `Matrix3`?
A long standing goal with the math library has been to fix all of the differences between the math types, and merging a math PR that makes this "worse" isn't great. I'm willing to spend some of my time too making the interfaces more uniform between the different math types.

vovatrykoz commented:
Sure, I can do that! Just to check my understanding, when you talk about math type consistency, do you mean making sure all matrices with the same dimensions can use the same operations? So, for example, all Vector2 types (Vector2, Vector2d, Vector2i etc) should have the same operators/functions/interfaces implemented? I also assume that the types working with half-prescision floats are an exception, since it is not possible to perform some of the mathematical operations on them

NogginBops commented:
Yeah that is basically what I'm talking about. And yeah Half is an exception. 
There are a few other exceptions that are out there where for example some floating point operations might not make sense for integer vectors and vice versa.

But for all the cases where it makes sense I think all matrices should support a certain set of interfaces, all vectors (except half) should support a certain set of interfaces etc.

vovatrykoz commented:
Okay, I've just added the missing interfaces to Matrix3d. I will double-check all the other math types, but it looks like the rest of the math types are as consistent as possible at the moment.

Some notable exceptions are
- Half types, as discussed
- Operators for VectorN * MatrixNi multiplication, since there are no integer matrices
- Multiplication for different dimensions. For example, Matrix3 supports IMultiplyOperators<Matrix3, Matrix3x2, Matrix3x2> but Matrix4 doesn't support IMultiplyOperators<Matrix4, Matrix3x2, ?> due to the laws of matrix multiplication

NogginBops commented:
> Multiplication for different dimensions. For example, Matrix3 supports IMultiplyOperators<Matrix3, Matrix3x2, Matrix3x2> but Matrix4 doesn't support IMultiplyOperators<Matrix4, Matrix3x2, ?> due to the laws of matrix multiplication

Yes, but `Matrix4` should support `IMultiplyOperators<Matrix4, Matrix4x2, Matrix4x2>` right?

vovatrykoz commented:
Yes, you are totally right, and it should also support `IMultiplyOperators<Matrix4, Matrix4x3, Matrix4x3>`. This also applies for the `Matrix4d`. I'll get on that!

vovatrykoz commented:
Added the missing `Matrix4` and `Matrix4d` interfaces along with the required operators and multiplication methods

vovatrykoz commented:
Also, I noticed that there are unit tests for Matrix4. Since I added new operators and methods for this type, I could also add the corresponding unit tests there, if that is ok

NogginBops commented:
> Also, I noticed that there are unit tests for Matrix4. Since I added new operators and methods for this type, I could also add the corresponding unit tests there, if that is ok

Sounds good 👍

vovatrykoz commented:
Ok, I have added unit tests for the 2 new operators on Matrix4. This is the only matrix that has unit tests so far. I honestly don't mind adding unit tests for other matrix types, but I feel like it could be it's own PR with how many tests are still to be added

NogginBops commented:
> Ok, I have added unit tests for the 2 new operators on Matrix4. This is the only matrix that has unit tests so far. I honestly don't mind adding unit tests for other matrix types, but I feel like it could be it's own PR with how many tests are still to be added

Yeah our unit tests have quite a few issues tbh, but yeah I can agree that making a separate PR for that might be appropriate.

sora-tech commented:
> It is not possible to implement some interfaces due to constraint on TSelf: TSelf needs to implement that same mathematical interface.

I discussed this with @NogginBops and they were happy for the operator definitions to move to satisfy the `TSelf` requirement, e.g. the current Matrix2 * Vector2 operator is defined in Vector2 but can be moved to Matrix2 to satisfy the constraint and provide an implementation for the interface.

Any operator for types outside of the library will not be possible e.g. `float * Quarternion` but so long as the other interface is available then you can make the maths work.

vovatrykoz commented:
Ok, then if we are to move it, I assume that only the operator should be moved, right? For example, the current [`Matrix4 * Vector4`](https://github.com/vovatrykoz/opentk/blob/b834fcc01cc26c35bc70d8a39f21e528028b5939/src/OpenTK.Mathematics/Vector/Vector4.cs#L2100) is defined as:

```
        public static Vector4 operator *(Matrix4 mat, Vector4 vec)
        {
            TransformColumn(in mat, in vec, out Vector4 result);
            return result;
        }
```

If we were to move it into Matrix4, we can either delegate the operation to the Vector4 method like this
```
        public static Vector4 operator *(Matrix4 mat, Vector4 vec)
        {
            Vector4.TransformColumn(in mat, in vec, out Vector4 result);
            return result;
        }
```

Or we could move the [`TransformColumn`](https://github.com/vovatrykoz/opentk/blob/b834fcc01cc26c35bc70d8a39f21e528028b5939/src/OpenTK.Mathematics/Vector/Vector4.cs#L1016) method into the `Matrix4` struct as well, but this would be an API-breaking change, since the code that utilizes `Vector4.TransformColumn(in mat, in vec, out Vector4 result)` would no longer be valid.

So, would it ok to only move out the operators but keep the underlying functions inside their original structs (aka, keep `TransformColumn` inside `Vector` while moving out the operator)? I was thinking that maybe doing it this would make the Math library a bit less organised and messier, but maybe I am overthinking it?

NogginBops commented:
> So, would it ok to only move out the operators but keep the underlying functions inside their original structs (aka, keep TransformColumn inside Vector while moving out the operator)? I was thinking that maybe doing it this would make the Math library a bit less organised and messier, but maybe I am overthinking it?

This is fine for this PR, we can figure out of we want to move implementations of these methods elsewhere later.

vovatrykoz commented:
Moved some operator definitions to satisfy the ``TSelf`` requirement so that we can add generic math interfaces corresponding to those operators. In general they were

- Quaternion * Vector operators
- Quaterniond * Vectord operators
- Matrix * Vector operators (includes both float and double matrices and vec)

NogginBops commented:
@vovatrykoz Should we consider adding more interfaces like `ISignedNumber<TSelf>` and `IMinMaxValue<TSelf>` (maybe even `IFloatingPointConstants<TSelf>`...) in this PR? Or should we consider that for a later PR?

vovatrykoz commented:
@NogginBops Well, it depends a bit on how you guys want to do it and how the PR's are normally structured here. On one hand, this pull request is pretty large already. Adding more would also mean more lines of code to review in a single PR before merging. 

On the other hand, I personally don't mind adding more to this PR. Also, both the PR and the linked issue are called "Generic Math Interfaces" and not just "Operator interfaces", so maybe we can just continue doing it here and just add everything at once. 

Either way, I think it would be better to define a list of all interfaces that we want to add so that we know what is there left to do. From what I could see and also based on your suggestions the following interfaces could be relevant:
- [``ISignedNumber<TSelf>``](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.isignednumber-1)
- [``IAdditiveIdentity<TSelf,TResult>``](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.iadditiveidentity-2)
- [``IMinMaxValue<TSelf>``](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.iminmaxvalue-1)
- [``IMultiplicativeIdentity<TSelf,TResult>``](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.imultiplicativeidentity-2)
- [``IFloatingPointConstants<TSelf>``](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.ifloatingpointconstants-1?view=net-8.0)

I assume that ``IFloatingPointConstants<TSelf>`` is for MathHelper, right? Also, any other interfaces I am missing?

If you are ok with me continuing on this PR, I can continue adding these. Otherwise, I'll just open a new PR for the rest of the interfaces

NogginBops commented:
> I assume that IFloatingPointConstants<TSelf> is for MathHelper, right? Also, any other interfaces I am missing?

I was thinking we would add something like `Vector3.Pi` but thinking about it, it doesn't make a ton of sense. So we can skip `IFloatingPointConstants<TSelf>`.

I would be fine reviewing a large PR that adds all of the math interfaces. I could also contribute to this PR if you want me to.


It might be interesting also to consider making something like `IVectorFunctions<TSelf>` ourselves that make it possible to use generic maths for stuff like `.Length` etc. Because I don't think there is a standardized interface in .net for that yet... I'm actually quite interested in doing that myself so I might give it a crack.

Charanor commented:
Eagerly awaiting this so I can re-write all my second-order dynamics related stuff to be generic instead of having 8 copy+pasted variants :D

NogginBops commented:
> Eagerly awaiting this so I can re-write all my second-order dynamics related stuff to be generic instead of having 8 copy+pasted variants :D

Oo interesting. What is the reason you have 8 copies of the functions? For what types? 

It would actually be great if you could share what operations your need to be available through the generics, e.g. are you using any vector specific functions? Maybe you could even share some code, so we can ensure it's possible to write it in a generic way?

Charanor commented:
@NogginBops I have plenty of stuff that would benefit from this, for example my implementation of the smooth spring function for Game Programing Gems 4 looks like this:

```cs
private static dynamic SpringDynamic(dynamic from, dynamic to, ref dynamic vel, float smoothTime, float delta)
{
    // Game Programming Gems 4: 1.10
    var omega = 2f / MathF.Max(smoothTime, y: 0.001f); // So we don't divide by 0
    var x = omega * delta;
    var exp = 1f / (1f + x + 0.48f * MathF.Pow(x, y: 2) + 0.235f * MathF.Pow(x, y: 3));
    var change = from - to;
    var temp = (vel + change * omega) * delta;
    vel = (vel - temp * omega) * exp;
    return to + (change + temp) * exp;
}

// Public interfaces for all applicable types
public static float Spring(float from, float to, ref float vel, float smoothTime, float delta) => /* cast values to dynamic and call SpringDynamic */
public static Vector2 Spring(Vector2 from, Vector2 to, ref Vector2 vel, float smoothTime, float delta) => /* cast values to dynamic and call SpringDynamic */
public static Vector3 Spring(Vector3 from, Vector3 to, ref Vector3 vel, float smoothTime, float delta) => /* cast values to dynamic and call SpringDynamic */
... and more ...
```
For this I would prefer to declare a single method like so:
```cs
public static dynamic Spring<T>(T from, T to, ref T vel, float smoothTime, float delta) 
    where T : IAdditionOperators<T, T, T>, ISubtractionOperators<T, T, T>, IMultiplyOperators<T, float, T>
{
    // Game Programming Gems 4: 1.10
    var omega = 2f / MathF.Max(smoothTime, y: 0.001f); // So we don't divide by 0
    var x = omega * delta;
    var exp = 1f / (1f + x + 0.48f * MathF.Pow(x, y: 2) + 0.235f * MathF.Pow(x, y: 3));
    var change = from - to;
    var temp = (vel + change * omega) * delta;
    vel = (vel - temp * omega) * exp;
    return to + (change + temp) * exp;
}
```
Similarly I have an interpolation util that can be applied to scalars or vectors. Currently I have separate methods for each type but it would benefit from declaring a generic version:
```cs
public delegate float Interpolator([ValueRange(from: 0, to: 1)] float value);

private readonly Interpolator interpolator;

public Interpolation(Interpolator interpolator)
{
    this.interpolator = interpolator;
}

// Several identical methods for all Vector types
[Pure]
public float Apply(float start, float end, [ValueRange(from: 0, to: 1)] float alpha) => start + (end - start) * interpolator(alpha);

// I would prefer a single one like this:

[Pure]
public T Apply<T>(in T start, in T end, [ValueRange(from: 0, to: 1)] float alpha) 
    where T : IAdditionOperators<T, T, T>, ISubtractionOperators<T, T, T>, IMultiplyOperators<T, float, T>
    => start + (end - start) * interpolator(alpha);
```
The second order dynamics one is a bit big to post code but these interfaces would be used:
```cs
IAdditionOperators<T, T, T>
ISubtractionOperators<T, T, T>
IDivisionOperators<T, float, T>
IMultiplyOperators<T, float, T>
IAdditiveIdentity<T, T> // Identity would be Vector2.Zero, Vector3.Zero, etc.
```
I have not yet found a need for `IMultiplicativeIdentity` but I can imagine it being useful, returning `Vector2.One`, `Vector3.One`, etc.

vovatrykoz commented:
I have added the following interfaces to the math library:

Vectors: 
- ``IAdditiveIdentity``
- ``IMultiplicativeIdentity``
- ``IMinMaxValue``

Square matrices:
- ``IAdditiveIdentity``
- ``IMultiplicativeIdentity``

Other matrices:
- ``IAdditiveIdentity``

Quaternions:
- ``IAdditiveIdentity``
- ``IMultiplicativeIdentity``

Some notes:
- Technically speaking an identity matrix could work as a multiplicative identity for a non-square matrix. For example, multiplying a 2x3 matrix with a 3x3 identity matrix will produce a new 2x3 matrix equal to the first one. However, I am unsure if it makes sense mathematically (not sure if a matrix with different dimensions can be considered an identity) and if it is possible to do programmatically, as the identity has to be of type ``TSelf``. In the case of a 2x3 matrix, ``TSelf`` is a 2x3 matrix, making it impossible to use a 3x3 matrix there
- Not sure if adding ``IMinMaxValue`` to matrices makes sense, as there is no way to say if a matrix is "greater" than another matrix. With vectors, you could use length to determine if one vector is larger than the other. But there is no such thing for matrices. It is still possible to add an implementation that returns matrices filled with positive/negative infinity for floats and doubles and min/max values for ints, but I am unsure if this is a good approach.

Other than that, I think the generic math interface implementation is pretty much complete, and the one thing that we have left to consider is the ``IVectorFunctions<TSelf>`` interface. Candidates for inclusion in this interface would probably be

- ``Length``
- ``LengthSquared``
- ``Normalize()``
- ``DotProduct()``
- ``CrossProduct()``

Maybe even math operations could be added, like ``Add()``, ``Subtract()``, ``Scale(float scalar)``, ``Scale(TSelf scalar)``?

NogginBops reviewed (changes requested):
Sorry for following up on this PR so late.
I've pushed a little commit myself to this PR just implementing `IMultiplicativeIdentity` for non-square matrices, and `IUnaryPlusOperators` for vectors. So remeber to sync with those changes so you can avoid merge conflicts with your own branch.

This PR is generally great work, I did however notice some problems while reviewing this PR. The main issue relates to the non symmetry in matrix vector multiplication order, and that only one of the two multiplication orders have been converted properly. This is unfortunately a very easy mistake to make as the interface has a default implementation for the reverse multiplication order which works for scalars but does not work for matrices as left or right multiplication matters. So these need to be properly implemented for both left and right multiplication.

I also think we should remove the multiplication interfaces on quaternions for a similar reason, we only support `quat * vector` multiplication with quaternions. But the interface forces us to implement both, which we do not want to support.

vovatrykoz commented on src/OpenTK.Mathematics/Matrix/Matrix2.cs:
```diff
@@ -759,6 +780,19 @@ public static void Swizzle(in Matrix2 mat, int rowForRow0, int rowForRow1, out M
+        /// <param name="vec">The vector to transform.</param>
+        /// <returns>The transformed vector.</returns>
+        [Pure]
+        public static Vector2 operator *(Matrix2 mat, Vector2 vec)
+        {
+            Vector2.TransformColumn(in mat, in vec, out Vector2 result);
+            return result;
+        }
+
```
The `(Vector, Matrix)` version of the operator does actually exist, but it is implemented in the Vector files. This is because the IMultiplyOperators has the following recursive constraint:

`public interface IMultiplyOperators<TSelf,TOther,TResult> where TSelf : IMultiplyOperators<TSelf,TOther,TResult>`. 

So TSelf needs to recursively implement the `IMultiplyOperators<TSelf,TOther,TResult>` interface. 

For example, if I remove the `IMultiplyOperators<Vector3, Matrix3, Vector3>` from Vector3 and define this interface on Matrix3 instead, for example:
`public struct Matrix3: IMultiplyOperators<Vector3, Matrix3, Vector3>` I will get the following error, since the Vector struct would not support the IMultiplyOperators interface:

`The type 'OpenTK.Mathematics.Vector3' cannot be used as type parameter 'TSelf' in the generic type or method 'IMultiplyOperators<TSelf, TOther, TResult>'. There is no boxing conversion from 'OpenTK.Mathematics.Vector3' to 'System.Numerics.IMultiplyOperators<OpenTK.Mathematics.Vector3, OpenTK.Mathematics.Matrix3, OpenTK.Mathematics.Vector3>'.[CS0315](https://msdn.microsoft.com/query/roslyn.query?appId=roslyn&k=k(CS0315))`

However, if I define the interface on both Matrix3 and Vector3 like so

`public struct Matrix3: IMultiplyOperators<Vector3, Matrix3, Vector3>`
`public struct Vector3: IMultiplyOperators<Vector3, Matrix3, Vector3>`

The compiler will let this pass (with a warning) as long as both types implement the required operator. However, now we have two operators with exact same signature on two different types, os doing something like this:

`
Vector3 vec3 = Vector3.One;
Matrix3 mat3 = Matrix3.Identity;
var example4 = vec3 * mat3;
`
will result in the following error, because both structs have the same operator with the exact same signature:
`Operator '*' is ambiguous on operands of type 'Vector3' and 'Matrix3'[CS0034](https://msdn.microsoft.com/query/roslyn.query?appId=roslyn&k=k(CS0034))`

For these reasons, the [`IMultiplyOperators<Matrix3, Vector3, Vector3>`](https://github.com/vovatrykoz/opentk/blob/600d6905e061c9690986a0f0f30fc81be9cd4e03/src/OpenTK.Mathematics/Matrix/Matrix3.cs#L39)
and the corresponding operator[ `public static Vector3 operator *(Matrix3 mat, Vector3 vec)`](https://github.com/vovatrykoz/opentk/blob/600d6905e061c9690986a0f0f30fc81be9cd4e03/src/OpenTK.Mathematics/Matrix/Matrix3.cs#L1292) 
is defined in Matrix3.cs. 

The [`IMultiplyOperators<Vector3, Matrix3, Vector3>`](https://github.com/vovatrykoz/opentk/blob/600d6905e061c9690986a0f0f30fc81be9cd4e03/src/OpenTK.Mathematics/Vector/Vector3.cs#L47)
and the corresponding operator [`public static Vector3 operator *(Vector3 vec, Matrix3 mat)`](https://github.com/vovatrykoz/opentk/blob/600d6905e061c9690986a0f0f30fc81be9cd4e03/src/OpenTK.Mathematics/Vector/Vector3.cs#L1657)
is defined inside Vector3.cs

Same for other dimensions

vovatrykoz commented on src/OpenTK.Mathematics/Data/Quaternion.cs:
```diff
@@ -32,7 +33,17 @@ namespace OpenTK.Mathematics
+                                ISubtractionOperators<Quaternion, Quaternion, Quaternion>,
+                                IMultiplyOperators<Quaternion, Quaternion, Quaternion>,
+                                IMultiplyOperators<Quaternion, float, Quaternion>,
+                                IMultiplyOperators<Quaternion, Vector2, Vector2>,
+                                IMultiplyOperators<Quaternion, Vector3, Vector3>,
+                                IMultiplyOperators<Quaternion, Vector4, Vector4>,
```
I am not sure it does. 
The interface `IMultiplyOperators<Quaternion, Vector2, Vector2>` requires that there is an operator with the signature `public static Vector2 operator *(Quaternion quat, Vector2 vec)`. 

It does not require the operator `public static Vector2 operator *(Vector2 vec, Quaternion quat)`. 
For that to be required, the interface should look like this: `IMultiplyOperators<Vector2, Quaternion, Vector2>`. No such operator is implemented on either quaternion types or any of the Vector types. I tried doing `vector * quat` just in case in my test project, and the compiler throws an error if I do that, but not if I do `quat * matrix`

vovatrykoz replied:
Hm, I took a look myself, and I coudn't find that. From what I could see, there is the usual static operator:

`static abstract TResult operator *(TSelf left, TOther right);`

and there is also a default implementation of the checked version of the same operator, which preserves the order and defauts to the unchecked operator

`static virtual TResult operator checked *(TSelf left, TOther right) => left * right;`

vovatrykoz commented:
Hey, thanks for the feedback! I responded to your comments and have created a new repo with a test project to demonstrate some of the things I described in the comments above. Can be found [here](https://github.com/vovatrykoz/test)

I think the main point is if there is an interface of type `IMultiplyOperators<A, B, C>`, it only requires implementing operator of type `A * B = C`, but not `B * A = C`. For that to be required you need to have `IMultiplyOperators<B, A, C>`. I understand this example might be a bit abstract, but I’ve provided more concrete examples in my comments on your revision requests.

Please feel free to reach out if anything is unclear or if you think I may have missed something! I believe I’ve addressed your concerns, but if there’s anything I’ve overlooked or misunderstood, I’m happy to revisit it! :)

NogginBops commented:
@vovatrykoz thank you for taking your time to make an example repo. When I have the time (tomorrow or the day after) I'll take a look at how it all works concretely. I just saw what visual studio told me was the interface definition but maybe I misunderstood. I'll make sure to thoroughly look through how the interfaces work to make sure I understand them correctly. Thank you for your patience with me on this! 🙂

vovatrykoz commented:
I also thought about identity matrices for non-square matrix types. In my understanding, an identity matrix acts like the number "1" during matrix multiplication. So given a random matrix `A` and an identity matrix `I`, we would get `A * I = A`. This is a bit tricky for non-square matrices, as they cannot be multiplied with each other due to row-column mismatch (for example, a 3x4 matrix cannot be multiplied with another 3x4 matrix, but a 3x4 matrix can be multiplied with a 4x3 matrix). 

So then there is a question of what do the identity matrices represent in this case. Is there, for example, a 4x3 matrix that can be used as an identity matrix for a 3x4 matrix? Should `Matrix3x4.Identity` produce a 4x3 matrix then? Or should `Matrix4x3.Identity` produce a matrix that would be an identity matrix for a 3x4 matrix? What about all other 4xD matrices? Is it even mathematically possible to have an identity matrix for a non-square matrix?

I also noticed that resources like [Wikipedia](https://en.wikipedia.org/wiki/Identity_matrix), and the book "Basic Linear Algebra" by Hillevi Gavel that I happen to own (written in Swedish, "Grundläggande linjär algebra") all define the identity matrix as a "square matrix" specifically, among its other properties. This got me wondering if there is a general way to define something similar for non-square matrices.

The book gives an example where a 3x3 identity matrix is multiplied with a 3x2 matrix,

|  |  |  |
|----------|----------|----------|  
| 1 | 0 | 0 |
| 0 | 1 | 0 |
| 0 | 0 | 1 |

multiplied by

|  |  |
|----------|----------|
| 1 | 2 |
| 3 | 4 |
| 5 | 6 |

which results in

|  |  |
|----------|----------|
| 1 | 2 |
| 3 | 4 |
| 5 | 6 |

And another example where a 2x2 identity matrix is multiplied with a 2x3 matrix. So it seems like identity matrices should be square matrices even for non-square matrices

|  |  |
|----------|----------|
| 1 | 0 | 
| 0 | 1 |

multiplied by

|  |  |
|----------|----------|
| -1 | -4 | 1 |
| 4 | -2 | -1 |

which results in

|  |  |
|----------|----------|
| -1 | -4 | 1 |
| 4 | -2 | -1 |

Both examples are taken from "Grundläggande linjär algebra" by Hillevi Gavel

Based on this I think we need to adjust the definition of the Identity matrix for non-square matrices

NogginBops commented:
@vovatrykoz I see what you are saying about with the identify matrices. I think if you make the changes you see fit with that (might be just removing the non square multiplicative identities) then I think that will be good.

Ps. Roligt att du kan svenska 🙂

vovatrykoz commented:
I removed all the non-square identity matrices, but I kept the IMultiplicativeIdentity interface on those matrices. I did it because, as demonstrated a 3x3 identity matrix also works for a 3x2 and a 3x4 matrices (and any other matrix with 3 rows). A 2x2 Identity matrix also works for a 2x3 and a 2x4 matrices.

For example, here is how the corresponding implementation looks for Matrix2x3:
The interface:
`public struct Matrix2x3: IMultiplicativeIdentity<Matrix2x3, Matrix3>`

The property
`public static Matrix3 MultiplicativeIdentity => Matrix3.Identity;`

Of course, the communtative property `A * I = I * A` is lost for non-square matrices, but I don't think this is an issue from what I've seen. People still use the square identity matrices as a multiplicative identity for the non-square ones.

In summary, while the concept of non-square identity matrices might not exist, the idea of multiplicative identifies for non-square matrices (via square matrices) does, so that is why I decided to keep the IMultiplicativeIdentity but remove the non-square identity matrices

> Ps. Roligt att du kan svenska 🙂

Jag har bott i Sverige i nästan 9 år nu). Har haft mycket tid på mig för att lära mig språket 😅

vovatrykoz commented:
A day later I realized that the things that I've described above with regards to `IMultiplicativeIdentity` is not an ideal solution. The thing is non-square matrices essentially have two different identity matrices: a left one and a right one. The left identity matrix has the same number of columns as the number of rows in a non-square matrix. The right identity matrix has the same number of rows as the number of columns in a non-square matrix.

For example, here is a 3x2 matrix has a 3x3 left identity matrix:
|  |  |  |
|----------|----------|----------|  
| 1 | 0 | 0 |
| 0 | 1 | 0 |
| 0 | 0 | 1 |

X

|  |  |
|----------|----------|
| 1 | 2 |
| 3 | 4 |
| 5 | 6 |

equals

|  |  |
|----------|----------|
| 1 | 2 |
| 3 | 4 |
| 5 | 6 |

And it also has a 2x2 right identity matrix:

|  |  |
|----------|----------|
| 1 | 2 |
| 3 | 4 |
| 5 | 6 |

X

|  |  |
|----------|----------|
| 1 | 0 | 
| 0 | 1 |

equals 

|  |  |
|----------|----------|
| 1 | 2 |
| 3 | 4 |
| 5 | 6 |

In code, this would look something like this:

`var example1 = Matrix3.Identity * mat3x2`
`var example2 = mat3x2 * Matrix2.Identity`

Both `example1` and `example2` are equal to `mat3x2` in this case


`IMultiplicativeIdentity` only allows defining one specific `MultiplicativeIdentity` property. My implementation currently returns a left identity matrix for non-square matrices. However, you could also return right identity matrix, and it would be equally correct. 

In theory, we could implement both versions of the interface. For example, it is possible to define both `IMultiplicativeIdentity<Matrix3x2, Matrix2>` and `IMultiplicativeIdentity<Matrix3x2, Matrix3>` with explicit interface implementation, which would in turn require some sort of upcasting to know which method to use, similar to how it works in F#. It would look something like this

```
public struct Matrix3x2 : IMultiplicativeIdentity<Matrix3x2, Matrix2>, IMultiplicativeIdentity<Matrix3x2, Matrix3>
{
    //...
    static Matrix2 IMultiplicativeIdentity<Matrix3x2, Matrix2>.MultiplicativeIdentity => Matrix2.Identity;
    static Matrix3 IMultiplicativeIdentity<Matrix3x2, Matrix3>.MultiplicativeIdentity => Matrix3Identity;
    // ...
}
```
Unfortunatelly, there is no easy way to [access static interface member when the interface is implemented explicitly](https://stackoverflow.com/questions/75286856/how-to-access-static-interface-member-when-the-interface-is-implemented-explicit). Accessing static members in this case requires some trickery and is quite awkward

I think a better approach in this case might be to not try to and "jam a square peg into a round hole" and simply remove the IMultiplicativeIdentity interface from the non-square matrices. Instead, I think we should define our own identity properties for non-square matrices: `LeftIdentity` and `RightIdentity`. This would look something like this
```
public struct Matrix3x2
{
    //...
    public static Matrix3 LeftIdentity => Matrix3.Identity;
    public static Matrix2 RightIdentity => Matrix2.Identity;
    // ...
}
```
this would allow you to do the following, given some `mat3x2` of type `Matrix3x2`:

`var example1 = Matrix3x2.LeftIdentity * mat3x2`
`var example2 = mat3x2 * Matrix3x2.RightIdentity`
Both `example1` and `example2` will be equal to `mat3x2` in this case.

I will make these changes (most likely) later today. I will remove the IMultiplicativeIdentity for the non-square matrices and just add `LeftIdentity` and `RightIdentity` properties to them. I feel like it would be a good enhancement for the non-square matrices.

NogginBops commented:
I don't think we need to add left and right identities to the library. We'll just skip identities for non square matrices.

vovatrykoz commented:
Ok, @NogginBops, I removed all multiplicative identities and the interfaces from the non-square matrices

vovatrykoz commented:
I just noticed something strange when reviewing the files. Certain non-square matrices have operators for multiplication with another matrix with the same dimension. For example in [Matrix3x4](https://github.com/vovatrykoz/opentk/blob/87160c9138fcf993bddb7123a96c22d1873bf367/src/OpenTK.Mathematics/Matrix/Matrix3x4.cs#L1050). The actual calculation is done in this [function](https://github.com/vovatrykoz/opentk/blob/87160c9138fcf993bddb7123a96c22d1873bf367/src/OpenTK.Mathematics/Matrix/Matrix3x4.cs#L804). It's been sitting there since before this pull request. When I was adding the interfaces I was adding it mechanically. However, reviewing this now I realized that was strange. I am no big linear algebra expert, so maybe this is some special multiplication algorithm for multiplying two non-square matrices of the same size?

NogginBops commented:
Regarding the non square matrix multiplication with matrices that don't technically match.
These are there as a convenience, for example it typical to store transforms in 4x3 matrices as the last column is usually just `(0, 0, 0, 1)`, but it would be annoying to work with if you couldn't multiply two transformations together so 4x3 x 4x3 matrix multiplication becomes the result of first extending the matrices to 4x4 (assuming the last column is `0, 0, 0, 1`), multiplying them and then discarding the `0, 0, 0, 1` column at the end. You will likely see this in other math libraries as well. Don't know if the documentation on these operations accurately describe what the function actually does though.

Regarding completing the rest of the matrix multiplication combinations, I would not have a problem if they are a part of this PR or if you want to make a new one. I'll leave that to your preference.

I want to once again thank you for the work on this PR, it's fantastic.

vovatrykoz commented:
> I want to once again thank you for the work on this PR, it's fantastic.

Appreciate it! 🙂 
It is fun contributing to such a great and well-maintained project as OpenTK.

> Regarding completing the rest of the matrix multiplication combinations, I would not have a problem if they are a part of this PR or if you want to make a new one. I'll leave that to your preference.

Well, we already started fixing some inconsistencies in this PR, so we might as well finish it! I will implement the missing multiplication operators in the coming days. After that, the PR should finally be done!

vovatrykoz commented:
Hey @NogginBops, just got done with adding the missing operators from the spreadsheet on the remainder of the math types. Everything that was marked as not implemented there should be done.

I've allowed myself the liberty of writing unit tests for some of the missing operators to make sure that I write them correctly, as there are a lot of combinations and it is easy to make a mistake here and there. This required me to add some missing generators so that FsCheck is able to work and test correctly.

NogginBops reviewed (approved):
This PR is a work of pure beauty.
I was somewhat dreading the code review of this PR but the code is super consistent and the added tests are the cherry on top!
I'll wait for the CI to pass, then I'll merge this.

vovatrykoz commented:
Thanks for the kind words, I really appreciate it. I tried to keep things as clean and consistent as possible, both for the sake of the review, but also in case anyone else digs into the code later (and also for the sake of my own sanity, so that I don't get confused when implementing 10 different multiplilcation methods with slight variations between them). Glad to hear it paid off!

NogginBops commented:
CI passes, merging this now!
Ping me on discord if you want the contributor role!
